src/unstract/sdk/utils/callback_manager.py

```diff
@@ -61,29 +58,43 @@ def set_callback_manager(
             )
         """
 
-        if llm:
-            tokenizer = CallbackManager.get_tokenizer(llm)
-        elif embedding:
-            tokenizer = CallbackManager.get_tokenizer(embedding)
+        # Nothing to do if callback manager is already set for the instance
+        if (
+            model
+            and model.callback_manager
+            and len(model.callback_manager.handlers) > 0
+        ):
+            return
+
+        model.callback_manager = CallbackManager.get_callback_manager(
+            model, platform_api_key, kwargs
+        )
 
+    @staticmethod
+    def get_callback_manager(
+        model: Union[LLM, BaseEmbedding],
+        platform_api_key: str,
+        kwargs,
+    ) -> LlamaIndexCallbackManager:
+        tokenizer = CallbackManager.get_tokenizer(model)
         token_counter = TokenCountingHandler(tokenizer=tokenizer, verbose=True)
+        llm = None
+        embedding = None
+        if isinstance(model, LLM):
+            llm = model
+        elif isinstance(model, BaseEmbedding):
+            embedding = model
         usage_handler = UsageHandler(
             token_counter=token_counter,
             platform_api_key=platform_api_key,
             llm_model=llm,
             embed_model=embedding,
-            **kwargs,
+            kwargs=kwargs,
         )
 
         callback_manager: LlamaIndexCallbackManager = LlamaIndexCallbackManager(
             handlers=[token_counter, usage_handler]
         )
-
-        if llm is not None:
-            llm.callback_manager = callback_manager
-        if embedding is not None:
-            embedding.callback_manager = callback_manager
-
         return callback_manager
```

Comment:
@gaya3-zipstack Better to have a thin wrapper on top and not expose LI callback manager directly at all.